validation-gen: unions

jpbetz opened the pull request:
Add union generation
Update enum generation

Added "support variables" which both unions and enums use to initialize state (eg. a set of enum symbols).

jpbetz commented on pkg/apis/core/v1/zz_generated.validations.go:
```diff
@@ -284,6 +288,23 @@ func Validate_EndpointsList(in *v1.EndpointsList, fldPath *field.Path) (errs fie
+	return errs
+}
+
+func Validate_EnvVarSource(in *v1.EnvVarSource, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.Union(fldPath, in, map[any]any{
+		"resourceFieldRef": []any{in.ResourceFieldRef},
+		"configMapKeyRef":  []any{in.ConfigMapKeyRef},
+		"secretKeyRef":     []any{in.SecretKeyRef},
+		"fieldRef":         []any{in.FieldRef},
+	})...)
+	return errs
```
There's probably a better way to do this.  What I was looking for was a simple way to provide:

The values of the fields in a struct that were tagged with `+unionMember`, grouped by the unionMember those fields belong to (multiple fields can belong to the same union member, this all aligns with the Unions KEP).

jpbetz replied:
one possible alternative is to generate all of the union enforcing code rather than building a structure to pass to a validator function

jpbetz replied:
I'm currently favoring:

```go
var envVarSourceUnionMembership = validate.NewUnionMembership("fieldRef", "resourceFieldRef", "configMapKeyRef", "secretKeyRef")

func Validate_EnvVarSource(in *v1.EnvVarSource, fldPath *field.Path) (errs field.ErrorList) {
	errs = append(errs, validate.Union(fldPath, in, envVarSourceUnionMembership, in.FieldRef, in.ResourceFieldRef, in.ConfigMapKeyRef, in.SecretKeyRef)...)
	return errs
}
```

This complicates codegen significantly, but it's general purpose, and enums could also use it to generate a var holding the set of supported values.

jpbetz commented:
@thockin This feels feature complete to me.  It expands on some parts of the generator, so I don't want to rush the merge.

thockin reviewed (commented):
Would be nice if this split enums changes from union changes :)

Just looked at the enum parts for now:

Shouldn't enum constants be limited to those defined in the same package as the enum type?  I had not looked at this code before but it feels weiid to allow another package to define values for an enum and they will be accepted as long as that othr package is linked?

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go:
```diff
@@ -51,3 +58,140 @@ func Required[T comparable](fldPath *field.Path, value T) field.ErrorList {
+//
+// For example:
+//
+//	var abcUnionMembership := schema.NewUnionMembership("a", "b", "b")
```
a, b, b, -> a, b, c  ?

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go:
```diff
@@ -51,3 +58,140 @@ func Required[T comparable](fldPath *field.Path, value T) field.ErrorList {
+//
+//	var abcUnionMembership := schema.NewUnionMembership("a", "b", "b")
+//	func ValidateABC(fldPath, *field.Path, in *ABC) (errs fields.ErrorList) {
+//		errs = append(errs, Union(fldPath, in, abcUnionMembership, in.A, in.B, in.C)...)
```
Which field path do we complain about?  All of them?  The parent struct?  The discriminator?

jpbetz replied:
I've gone with the parent struct.  The validation of the discriminator is typically handled via enum validation of the discriminator (and is handled separate from the union validation).

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go:
```diff
@@ -51,3 +58,140 @@ func Required[T comparable](fldPath *field.Path, value T) field.ErrorList {
+//	}
+func Union(fldPath *field.Path, in any, union *UnionMembership, fieldValues ...any) field.ErrorList {
+	if len(union.members) != len(fieldValues) {
+		return field.ErrorList{field.InternalError(fldPath, fmt.Errorf("unexpected difference in length between fields defined in UnionMembership as fieldValues"))}
```
s/as/and ?

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go:
```diff
@@ -51,3 +58,140 @@ func Required[T comparable](fldPath *field.Path, value T) field.ErrorList {
+		if rv.IsValid() && !rv.IsZero() {
+			m := union.members[i]
+			if specifiedMember != nil && *specifiedMember != m.memberName {
+				return field.ErrorList{field.Invalid(fldPath, "", fmt.Sprintf("must not specify %s, must specify one of: %s", union.specifiedFields(fieldValues), union.allFields()))}
```
As we reach completion of this, I'd like to think hard about the most useful phrasing of the error message.

jpbetz replied:
Please check me on this.  I've made an effort to produce errors that focus on what fields should be specified, e.g.:

`must not specify "b" and "d", must specify one of: "a", "b", "c" or "d"` (non-discriminated)

`"b" must not be specified when "type" is "C"` (discriminated)

The major alternative I considered was to include the concept of "union" directly in the error messages. Happy to consider phrasing taking that approach if it helps with clarity.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/zz_generated.validations.go:
```diff
@@ -0,0 +1,240 @@
+	return errs
+}
+
+var u1UnionMembership = validate.NewUnionMembership("m1", "m2")
```
As with enums, I wonder if `sync.Once()` would let us move this inside the function

thockin replied:
This was a stupid suggestion

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/zz_generated.validations.go:
```diff
@@ -0,0 +1,240 @@
+var u1UnionMembership = validate.NewUnionMembership("m1", "m2")
+
+func Validate_U1(in *U1, fldPath *field.Path) (errs field.ErrorList) {
+	errs = append(errs, validate.Union(fldPath, in, u1UnionMembership, in.M1, in.M2)...)
```
So this reports an error on the containing struct, which really precludes having 2 unions in a struct, right?

Shouldn't this be fatal wrt the union members?  E.g. 

```
if e := validate.Union(fldPath, in, u1UnionMembership, in.M1, in.M2); len(e) != 0 {
    errs = append(errs, e...)
} else {
    // validate M1
    // validate M2
}
```

I think the func() thing might make this easier?

thockin replied:
Thinking more about this, it's tricky.  How would we express this?

One idea before I forget it:
* add an optional "Dependants()" method to GenFunction.  It returns a list of fields whose validation can only proceed if *this* validation succeeds  
* union validator returns the union members from Dependants()
* code generates something like:

```
Validate_U1(...) {
    failUnion<hash> := false
    if e := validate.Union(...); len(e) != 0 {
        errs = append(...
        failUnion<hash> = true
    }

    // M1
    if !failUnion<hash> {
        // other M1 validation
    }
```

thockin replied:
Trying to find other ways to approach this.  Thinking out loud.

Maybe make less scaffolding for generators.  Just pass them a SnippetWriter and maybe an Args, and let them emit code with a given assumed environment.

E.g.  Instead of calling `v.ExtractValidations()` and then iterating, just call `v.Render(sw)`.  Each validator can assume:
* `obj` is the object in question
* `fldPath` is the current field path
* `errs` is where to append results
* they can `return` at will, and bypass any further validation

The `required` validator would emit the `if e:= ...; len(e) != 0` logic, or maybe we keep the flags stuff to make it a bit easier.  Same for nil-checks.

The `union` validator... this doesn't help, does it?  We still need a file-scope variable and we still need to influence OTHER field validations.  I'll keep thinking about it.  I am pretty much out of time to review further, but I may yet bring laptop on vacation because this is fun.

jpbetz replied:
Oh boy, yes. This is a big gap. Looking more closely.

jpbetz replied:
I'll hold on this until dependent fields is sorted out.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/enums/zz_generated.validations.go:
```diff
@@ -0,0 +1,111 @@
+	return nil
+}
+
+var e1Symbols = sets.New[E1]()
```
Good catch. I'll switch to the prefix approach and add a TODO for the hash.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/enums/doc.go:
```diff
@@ -0,0 +1,65 @@
+type E4 string
+
+const (
+	E4S1 E4 = "customname"
```
I'll add a comment.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/enums/zz_generated.validations.go:
```diff
@@ -0,0 +1,111 @@
+
+	// E1
+	errs = append(errs, validate.FixedResult(fldPath.Child("e1"), in.E1, true, "field T1.E1")...)
+	errs = append(errs, Validate_E1(&in.E1, fldPath.Child("e1"))...)
```
Fixing this to call fatal before non-fatal does seem "more correct".   Should we wait till we have a concrete use case?

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go:
```diff
@@ -33,11 +39,12 @@ func MaxLength(fldPath *field.Path, value string, max int) field.ErrorList {
-	valueString := string(value)
-	if !slices.Contains(symbols, valueString) {
-		return field.ErrorList{field.NotSupported(fldPath, value, symbols)}
+func Enum[T ~string](fldPath *field.Path, value *T, symbols sets.Set[T]) field.ErrorList {
```
Good call.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/enum.go:
```diff
@@ -42,27 +46,39 @@ type enumDeclarativeValidator struct {
+	if _, ok := commentTags[tagEnumType]; ok {
+		var result []FunctionGen
+		if enum, ok := c.enumContext.EnumType(t); ok {
+			supportVarName := PrivateVar{Name: t.Name.Name + "Symbols", Package: "local"}
```
There is code that blows up if the package is `""`, but if it's non-empty everything works because the package name never gets used for anything...

jpbetz commented:
Quick update on this one:
- I've rebased and moved all the enum support out (I'll submit a separate PR with the feedback from this PR incorporated)
- https://github.com/jpbetz/kubernetes/pull/11#discussion_r1698966035 will need to be carefully addressed.  I'm working on that.
- I need to restructure `ExtractValidations` to return a struct that contains a list of `FunctionGen` and a list of `SupportVarGen`

jpbetz commented:
@thockin OK, this is ready for more review.  The only major gap is fatal/dependent field support (per offline discussion-- I'll hold on that since you have it already in progress).

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/enums/doc.go:
```diff
@@ -0,0 +1,65 @@
+	// +required
+	EP2 *E2 `json:"ep2"`
+
+	E3  E3  `json:"e3"`
```
I've simplified this down to just enums with 0,1,2 values, with and without pointers.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/annotations/jsontagutil.go:
```diff
@@ -28,27 +28,27 @@ import (
 }
 
-func lookupJSONTags(m types.Member) (JSONTags, bool) {
+func LookupJSONTags(m types.Member) (JSONTags, bool) {
```
SGTM. https://github.com/kubernetes/gengo/pull/273

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -1067,21 +1121,58 @@ func emitCallsToValidators(c *generator.Context, validations []validators.Functi
+			sw.Do("}", nil)
+		default:
+			// TODO: check this during discovery and emit an error with more useful information
+			klog.Fatalf("Unsupported extraArg type: %T", value)
```
I thought I fixed all these, but this should be: `panic(fmt.Sprintf(...))`

jpbetz replied:
My bad. Fixing all occurrences I've introduced.

thockin reviewed (commented):
For enums: Can you add an output_test which exercises the options (no values, 1 value, N values) and tag?

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/validators.go:
```diff
@@ -61,18 +88,46 @@ type FunctionGen interface {
 	// a function that does and use that function to call the validation function.
 	SignatureAndArgs() (function types.Name, extraArgs []any)
 
+	TypeArgs() []types.Name
```
You got it! I'll add godoc explaining.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/validators.go:
```diff
@@ -61,18 +88,46 @@ type FunctionGen interface {
+// Var is a variable name that the generator will output as an identifier.
+type Var types.Name
+
+// PrivateVar is a variable name that the generator will output as a private identifier.
```
Yeah, this is for a PublicVar. I can't remember why I included this.  I don't use it for anything. I'll drop it.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/enum.go:
```diff
@@ -41,10 +41,10 @@ type enumDeclarativeValidator struct {
-	var result []FunctionGen
-	if enum, ok := c.enumContext.EnumType(t); ok {
-		result = append(result, Function("enum", DefaultFlags, enumValidator, enum.ValueArgs()...))
+func (c *enumDeclarativeValidator) ExtractValidations(t *types.Type, comments []string) (ValidatorGen, error) {
```
openapi-gen currently expects it be on the type.  This case is notable: https://github.com/kubernetes/kubernetes/blob/18b81ab68ca3238f05c59f0c63f501166a1fa843/staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go#L29-L30

I've split the enums adoption of support vars into https://github.com/jpbetz/kubernetes/pull/22.  Okay if I add more tests there?

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -164,6 +167,40 @@ func (g *genValidations) hasValidationsMiss(n *typeNode) bool {
 }
 
 // typeDiscoverer contains fields necessary to build graphs of types.
+func (g *genValidations) emitValidationVariables(c *generator.Context, t *types.Type, sw *generator.SnippetWriter) {
```
> also your merge messed up the godoc comment here :)

I had that coming :)

Fixed and added more godoc since it was lacking.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -164,6 +167,40 @@ func (g *genValidations) hasValidationsMiss(n *typeNode) bool {
+		if len(typeArgs) > 0 {
+			sw.Do("[", nil)
+			for i, typeArg := range typeArgs {
+				sw.Do("$.|raw$", c.Universe.Type(typeArg))
```
I needed this in enum support vars for the empty case (`sets.New[EnumType]()`) since inference doesn't help there.  I've added a test.

thockin replied:
`sets.New(values...)` would invoke type inference (assuming the values were appropriately typed as the enum type?

I'm not against keeping type-args here, just saying :)

jpbetz replied:
Ah, you're saying I could do `sets.New([]Type{}...)`.  to handle the empty case? Good point.  I'd be okay with that approach too.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -807,7 +848,7 @@ func (g *genValidations) emitValidationForChild(c *generator.Context, thisChild
 			// turns out to be a real problem, we could change this to pass
 			// everything by pointer.
-			emitCallsToValidators(c, validations, elemIsPtr, elemSW)
+			emitCallsToValidators(c, validations.Functions, elemIsPtr, elemSW)
```
Really good catch.  I'll add the error.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -1042,9 +1083,22 @@ func emitCallsToValidators(c *generator.Context, validations []validators.Functi
+			if len(typeArgs) > 0 {
+				sw.Do("[", nil)
+				for i, typeArg := range typeArgs {
+					sw.Do("$.|raw$", c.Universe.Type(typeArg))
```
I've always found `raw` to be a confusing term, but it seems to behave right--  it ensures the type's package is imported and then references the type.  Added a test for function calls.  https://github.com/jpbetz/kubernetes/pull/22 also shows it in use for support vars.

thockin replied:
ACK

jpbetz commented:
Thanks @thockin!  Feedback addressed.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/testing.go:
```diff
@@ -123,11 +132,19 @@ func (_ fixedResultDeclarativeValidator) parseTagVal(in string) (FunctionFlags,
+	var typeArgs []types.Name
+	if len(pl.TypeArg) > 0 {
+		index := strings.LastIndex(pl.TypeArg, ".")
+		if index < 0 {
```
Can the type-arg be `string` ?  Won't that fail here?

jpbetz replied:
Good catch. Fixed.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/type_args/doc.go:
```diff
@@ -0,0 +1,30 @@
+type T1 struct {
+	TypeMeta int
+
+	// +validateTrue={"type_arg": "k8s.io/code-generator/cmd/validation-gen/output_tests/primitives.T1", "msg": "T1.S"}
```
Add a case for a local type (`type E1 string`) and for a primitive type (`int`) ?

thockin reviewed (commented):
wrt enums, will look at unions next

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -460,6 +460,9 @@ func (td *typeDiscoverer) discoverStruct(thisNode *typeNode, fldPath *field.Path
 			if !validations.Empty() {
 				klog.V(5).InfoS("found field-attached validations", "n", validations.Len())
 				child.fieldValidations.Add(validations)
+				if len(validations.Variables) > 0 {
```
How about when we find list and map key/val validations?

jpbetz replied:
oops, added

jpbetz commented:
Thanks @thockin. Feedback applied

thockin reviewed (commented):
This covers all the union commits and the most recent updated ones.  Feel free to reorganize commits and squash as needed.  Appreciate the separation of enum and union

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/union.go:
```diff
@@ -0,0 +1,165 @@
+			}
+			tag := tags[0]
+			var fieldName string
+			if jsonAnnotation, ok := annotations.LookupJSONTags(member); ok {
```
> Shouldn't we be vocal about the error if we find a field which claims to be a unionMember but doesn't have JSON tags or doesn't have a name?

Added

> Edit: We need JSON tags pretty much everywhere, maybe we should just have some blanket assertion that validator code will never even get called if that isn't true? Since you (corrently) removed the panic - where can we do that?

This seems appropriate. OK if I put this in the TODO as a follow up?

thockin replied:
TODO SGTM

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/union.go:
```diff
@@ -0,0 +1,165 @@
+const (
+	// +union and +unionDiscriminator tag are used by openapi-gen to publish x-kubernetes-union and x-kubernetes-discriminator
+	// extensions into Kubernetes published OpenAPI.
+	discriminatorTagName = "unionDiscriminator"
```
+100.  Just today someone asked on slack about the `listMapKey` tag and while we did have _something_ to point to, it highlighted that our docs were not in a great state. I'll add to the TODO.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/union.go:
```diff
@@ -0,0 +1,165 @@
+
+			p := &discriminatorParams{}
+			if len(tag) > 0 {
+				if err := json.Unmarshal([]byte(tag), &p); err != nil {
```
I like this idea in general.

My only hesitation with union name is that it's for a feature we expect to rarely use-- having multiple unions in the same go struct.  We don't have any APIs in core that need this..

thockin replied:
Can we make what `validateTrue` does be "normal"?  Start with no value: `+validateTrue`.  Eventually we want a string or some other value: `+validateTrue="optional value"`.  Eventually we want something more complex: `+validateTrue={"msg": "optional value", "flags": [ "flag", "flag" ]}`.

We can document that and I think it will be OK

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -0,0 +1,162 @@
+// For example:
+//
+//	var abcUnionMembership := schema.NewUnionMembership("a", "b", "c")
+//	func ValidateABC(fldPath, *field.Path, in *ABC) (errs fields.ErrorList) {
```
Good eye.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -0,0 +1,162 @@
+// NewUnionMembership returns a new UnionMembership for the given list of members.
+//
+// Each member entry may one of:
+//   - string: To provide the fieldName of a field union member where the union member name is CamelCase of the fieldName.
```
Sure, I'll just include the pair always.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -0,0 +1,162 @@
+		if rv.IsValid() && !rv.IsZero() {
+			m := union.members[i]
+			if specifiedMember != nil && *specifiedMember != m.memberName {
+				return field.ErrorList{field.Invalid(fldPath, "", fmt.Sprintf("must not specify %s, must specify one of: %s", union.specifiedFields(fieldValues), union.allFields()))}
```
Cool, I've tried to incorporate this.

jpbetz commented on pkg/apis/core/v1/zz_generated.validations.go:
```diff
@@ -464,13 +464,6 @@ func Validate_EndpointsList(obj *v1.EndpointsList, fldPath *field.Path) (errs fi
 }
 
 func Validate_EphemeralContainer(obj *v1.EphemeralContainer, fldPath *field.Path) (errs field.ErrorList) {
-	// field v1.EphemeralContainer.EphemeralContainerCommon
```
When I ported the JSON tag parsing and put it into gengo, I fixed a bug where it was defaulting the JSON name for tags like `json:",inline"` to the go struct field name.  As a result, our generator was treating this inline struct member as a normal field.  We probably do need to deal with inlined types so I left a comment about it (I see you found it below!)

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -0,0 +1,162 @@
+		if isSpecified && !isDiscriminatedMember {
+			errs = append(errs, field.Invalid(fldPath.Child(member.fieldName), "", fmt.Sprintf("must not be specified when %s is %s", strconv.Quote(fmt.Sprint(union.discriminatorName)), strconv.Quote(fmt.Sprint(discriminatorValue)))))
+		} else if !isSpecified && isDiscriminatedMember {
+			errs = append(errs, field.Invalid(fldPath.Child(member.fieldName), "", fmt.Sprintf("must be specified when %s is %s", strconv.Quote(fmt.Sprint(union.discriminatorName)), strconv.Quote(fmt.Sprint(discriminatorValue)))))
```
Nice. Fixed.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/doc.go:
```diff
@@ -0,0 +1,168 @@
+// +k8s:validation-gen=TypeMeta
+
+// This is a test package.
+package unions
```
Good thing I can edit fast!  I'll try to make a habit of it.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/validators.go:
```diff
@@ -21,9 +21,36 @@ import "k8s.io/gengo/v2/types"
+}
+
+// ValidatorGen defines the function calls and variables to generate to perform validation.
+type ValidatorGen struct {
```
Super nit - should this be called `Validations` now?  I think it would read better.

jpbetz replied:
SGTM!

thockin reviewed (commented):
Almost all my comments are very minor.  LGTM overall.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -0,0 +1,155 @@
+		if isSpecified && !isDiscriminatedMember {
+			errs = append(errs, field.Invalid(fldPath.Child(member.fieldName), "",
+				fmt.Sprintf("may only be specified when `%s` is %q",
+					fmt.Sprint(union.discriminatorName),
```
geez, how tired was I?  Fixed

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/union.go:
```diff
@@ -0,0 +1,155 @@
+// [0] identifies the field name and [1] identifies the union member Name.
+//
+// Field names must be unique.
+func NewUnionMembership(member ...any) *UnionMembership {
```
Fixed

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/discriminated_union/doc.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+// +validateTrue="type T1"
+type T1 struct {
```
Good point. Dropping

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/multiple_discriminated_unions/zz_generated.validations.go:
```diff
@@ -0,0 +1,174 @@
+func Validate_DU(obj *DU, fldPath *field.Path) (errs field.ErrorList) {
+	// type DU
+	if obj != nil {
+		errs = append(errs, validate.DiscriminatedUnion(fldPath, *obj, unionMembershipForDUunion2, obj.DU2, obj.U2M1, obj.U2M2)...)
```
As long as we can convince ourselves it is not going to oscillate between builds, maybe it's fine?

jpbetz replied:
I'll hold on this for now. I've got a snippet of a sorter that should do the trick, but it's a bigger change.

thockin replied:
ACK

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/discriminated_union_custom_members/doc.go:
```diff
@@ -0,0 +1,65 @@
+
+type D string
+
+const (
```
The const? I'd like to keep it so the enum usage is exemplary.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -163,7 +168,7 @@ func (g *genValidations) hasValidationsMiss(n *typeNode) bool {
 }
 
-// typeDiscoverer contains fields necessary to build graphs of types.
+// typeDiscoverer contains fields necessary to build a tree of types.
```
Can't remember why I did this edit. reverting.

jpbetz commented:
@thockin Feedback applied in latest commit
